ReactWindows/ReactNative/UIManager/UIViewOperationQueue.cs

```diff
@@ -36,19 +36,54 @@ private class QueueInstanceInfo
             public int rootViewCount;
         }
 
-        private UIViewOperationQueueInstance _mainUiViewOperationsQueueInstance;
+        // Thread safe dictionary using a single lock
+        private class SimpleLockedDictionary<TKey, TValue>
+        {
+            private readonly object _lock = new object();
+            private readonly Dictionary<TKey, TValue> _dict = new Dictionary<TKey, TValue>();
 
-        // Maps tags to corresponding UIViewOperationQueueInstance
-        private readonly IDictionary<int, UIViewOperationQueueInstance> _reactTagToOperationQueue = new Dictionary<int, UIViewOperationQueueInstance>();
+            public void Add(TKey key, TValue value)
+            {
+                lock(_lock)
+                {
+                    _dict.Add(key, value);
+                }
+            }
 
-        // Maps CoreDispatcher to corresponding UIViewOperationQueueInstance + rootView accounting
-        private readonly IDictionary<CoreDispatcher, QueueInstanceInfo> _dispatcherToOperationQueueInfo = new Dictionary<CoreDispatcher, QueueInstanceInfo>();
+            public bool TryGetValue(TKey key, out TValue value)
+            {
+                lock (_lock)
+                {
+                    return _dict.TryGetValue(key, out value);
+                }
+            }
+
+            public void RemoveList(IList<TKey> keyList)
+            {
+                lock (_lock)
+                {
+                    foreach (var key in keyList)
+                    {
+                        _dict.Remove(key);
+                    }
+                }
+            }
+        }
 
-        // Action queue used by UIManagerModule
-        private IActionQueue _uiManagerActionQueue;
+        // Maps tags to corresponding UIViewOperationQueueInstance.
+        // Uses a simple serialized Dictionary that depends on a single lock (rather than one per bucket), allows fast removal of list of keys, etc.
+        private readonly SimpleLockedDictionary<int, UIViewOperationQueueInstance> _reactTagToOperationQueue = new SimpleLockedDictionary<int, UIViewOperationQueueInstance>();
+
+        // The _dispatcherToOperationQueueInfo (all entries but the main window related one) & _active combo are protected by this lock
+        // This allows for consistent lifecycle OnSuspend/OnResume events to e sent to each queue instance
+        private readonly object _lock = new object();
 
         private bool _active;
 
+        // Maps CoreDispatcher to corresponding UIViewOperationQueueInstance + rootView accounting
+        // The concurrent dictionary allows for thread safe ".Values" calls, crash-safe enumerating, etc.
+        private readonly ConcurrentDictionary<CoreDispatcher, QueueInstanceInfo> _dispatcherToOperationQueueInfo = new ConcurrentDictionary<CoreDispatcher, QueueInstanceInfo>();
```

Comment:
>ConcurrentDictionary [](start = 25, length = 20)

Might be overkill, but you could also consider a vanilla Dictionary and a ReaderWriterLockSlim.